In [1]:
#Adagrad, Adam, Adamax, ASGD, LBFGS, RMSprop, Rprop, SGD
import torch
from torch.autograd import Variable

In [2]:
x_data = Variable(torch.Tensor([[1.0],[2.0],[3.0]]))
y_data = Variable(torch.Tensor([[2.0],[4.0],[6.0]]))

In [3]:
class Model(torch.nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.linear =torch.nn.Linear(1,1)
    def forward(self, x):
        y_pred = self.linear(x)
        return y_pred

In [4]:
adagrad_model = Model()
adam_model = Model()
adamax_model = Model()
asgd_model = Model()
lbfgs_model = Model()
rmsprop_model = Model()
rprop_model = Model()
sgd_model = Model()

In [5]:
#Adagrad
criterion = torch.nn.MSELoss(size_average=False)
optimizer= torch.optim.Adagrad(adagrad_model.parameters(), lr = 0.01)

for epoch in range(500):
    y_pred = adagrad_model(x_data)
    
    loss= criterion(y_pred, y_data)
    
    optimizer.zero_grad()
    loss.backward()
    optimizer.step() 
pred_val = adagrad_model.forward(Variable(torch.Tensor([[4.0]]))).data[0][0]
print(pred_val)

tensor(3.8546)


C:\ProgramData\Anaconda3\lib\site-packages\torch\nn\_reduction.py:44: UserWarning: size_average and reduce args will be deprecated, please use reduction='sum' instead.
  warnings.warn(warning.format(ret))


In [6]:
#Adam
criterion = torch.nn.MSELoss(size_average=False)
optimizer= torch.optim.Adam(adam_model.parameters(), lr = 0.01)

for epoch in range(500):
    y_pred = adam_model(x_data)
    
    loss= criterion(y_pred, y_data)
    
    optimizer.zero_grad()
    loss.backward()
    optimizer.step() 
pred_val = adam_model.forward(Variable(torch.Tensor([[4.0]]))).data[0][0]
print(pred_val)

tensor(7.5368)


In [7]:
#Adamax
criterion = torch.nn.MSELoss(size_average=False)
optimizer= torch.optim.Adamax(adamax_model.parameters(), lr = 0.01)

for epoch in range(500):
    y_pred = adamax_model(x_data)
    
    loss= criterion(y_pred, y_data)
    
    optimizer.zero_grad()
    loss.backward()
    optimizer.step() 
pred_val = adamax_model.forward(Variable(torch.Tensor([[4.0]]))).data[0][0]
print(pred_val)

tensor(6.5678)


In [8]:
#ASGD
criterion = torch.nn.MSELoss(size_average=False)
optimizer= torch.optim.ASGD(asgd_model.parameters(), lr = 0.01)

for epoch in range(500):
    y_pred = asgd_model(x_data)
    
    loss= criterion(y_pred, y_data)
    
    optimizer.zero_grad()
    loss.backward()
    optimizer.step() 
pred_val = asgd_model.forward(Variable(torch.Tensor([[4.0]]))).data[0][0]
print(pred_val)

tensor(7.9703)


In [9]:
#LBFGS
criterion = torch.nn.MSELoss(size_average=False)
optimizer= torch.optim.LBFGS(lbfgs_model.parameters(), lr = 0.01)

def closure():
    optimizer.zero_grad()
    y_pred = lbfgs_model(x_data)
    loss = criterion(y_pred, y_data)
    loss.backward()
    return loss

for epoch in range(500):
    
    optimizer.step(closure) 
pred_val = lbfgs_model.forward(Variable(torch.Tensor([[4.0]]))).data[0][0]
print(pred_val)

tensor(8.0000)


In [10]:
#RMSprop
criterion = torch.nn.MSELoss(size_average=False)
optimizer= torch.optim.RMSprop(rmsprop_model.parameters(), lr = 0.01)

for epoch in range(500):
    y_pred = rmsprop_model(x_data)
    
    loss= criterion(y_pred, y_data)
    
    optimizer.zero_grad()
    loss.backward()
    optimizer.step() 
pred_val = rmsprop_model.forward(Variable(torch.Tensor([[4.0]]))).data[0][0]
print(pred_val)

tensor(7.3124)


In [11]:
#Rprop
criterion = torch.nn.MSELoss(size_average=False)
optimizer= torch.optim.Rprop(rprop_model.parameters(), lr = 0.01)

for epoch in range(500):
    y_pred = rprop_model(x_data)
    
    loss= criterion(y_pred, y_data)
    
    optimizer.zero_grad()
    loss.backward()
    optimizer.step() 
pred_val = rprop_model.forward(Variable(torch.Tensor([[4.0]]))).data[0][0]
print(pred_val)

tensor(8.0000)


In [12]:
#SGD
criterion = torch.nn.MSELoss(size_average=False)
optimizer= torch.optim.SGD(sgd_model.parameters(), lr = 0.01)

for epoch in range(500):
    y_pred = sgd_model(x_data)
    
    loss= criterion(y_pred, y_data)
    
    optimizer.zero_grad()
    loss.backward()
    optimizer.step() 
pred_val = sgd_model.forward(Variable(torch.Tensor([[4.0]]))).data[0][0]
print(pred_val)

tensor(7.9744)


Adagrad : 3.8546
Adam : 7.5368
Adamax : 6.5678
ASGD : 7.9703
LBFGS : 8.0000
RMSprop : 7.3124
Rprop : 8.0000
SGD : 7.9744